controlplane/kubeadm/internal/workload_cluster_coredns_test.go

```diff
@@ -29,11 +29,253 @@ import (
+						ClusterConfiguration: &kubeadmv1.ClusterConfiguration{
+							DNS: kubeadmv1.DNS{
+								Type: "foobarDNS",
+							},
+						},
+					},
+				},
+			},
+			objs:      []runtime.Object{badCM},
+			expectErr: false,
+		},
+		{
+			name:      "returns early without error if CoreDNS info is not found",
+			kcp:       validKCP,
+			expectErr: false,
+		},
+		{
+			name:      "returns error if there was a problem retrieving CoreDNS info",
+			kcp:       validKCP,
+			objs:      []runtime.Object{badCM},
+			expectErr: true,
+		},
+		{
+			name:      "returns early without error if CoreDNS fromImage == ToImage",
+			kcp:       validKCP,
+			objs:      []runtime.Object{depl, cm},
+			expectErr: false,
+		},
+		{
+			name: "returns error if validation of CoreDNS image tag fails",
+			kcp: &controlplanev1.KubeadmControlPlane{
+				Spec: controlplanev1.KubeadmControlPlaneSpec{
+					KubeadmConfigSpec: cabpkv1.KubeadmConfigSpec{
+						ClusterConfiguration: &kubeadmv1.ClusterConfiguration{
+							DNS: kubeadmv1.DNS{
+								Type: kubeadmv1.CoreDNS,
+								ImageMeta: kubeadmv1.ImageMeta{
+									// image is older than what's already
+									// installed.
+									ImageRepository: "k8s.gcr.io/some-folder/coredns",
+									ImageTag:        "1.1.2",
+								},
+							},
+						},
+					},
+				},
+			},
+			objs:      []runtime.Object{depl, cm},
+			expectErr: true,
+		},
+		{
+			name: "returns error if unable to update CoreDNS image info in kubeadm config map",
+			kcp: &controlplanev1.KubeadmControlPlane{
+				Spec: controlplanev1.KubeadmControlPlaneSpec{
+					KubeadmConfigSpec: cabpkv1.KubeadmConfigSpec{
+						ClusterConfiguration: &kubeadmv1.ClusterConfiguration{
+							DNS: kubeadmv1.DNS{
+								Type: kubeadmv1.CoreDNS,
+								ImageMeta: kubeadmv1.ImageMeta{
+									// provide an newer image to update to
+									ImageRepository: "k8s.gcr.io/some-folder/coredns",
+									ImageTag:        "1.7.2",
+								},
+							},
+						},
+					},
+				},
+			},
+			// no kubeadmConfigMap available so it will trigger an error
+			objs:      []runtime.Object{depl, cm},
+			expectErr: true,
+		},
+		{
+			name: "returns error if unable to update CoreDNS corefile",
+			kcp: &controlplanev1.KubeadmControlPlane{
+				Spec: controlplanev1.KubeadmControlPlaneSpec{
+					KubeadmConfigSpec: cabpkv1.KubeadmConfigSpec{
+						ClusterConfiguration: &kubeadmv1.ClusterConfiguration{
+							DNS: kubeadmv1.DNS{
+								Type: kubeadmv1.CoreDNS,
+								ImageMeta: kubeadmv1.ImageMeta{
+									// provide an newer image to update to
+									ImageRepository: "k8s.gcr.io/some-folder/coredns",
+									ImageTag:        "1.7.2",
+								},
+							},
+						},
+					},
+				},
+			},
+			migrator: &fakeMigrator{
+				migrateErr: errors.New("failed to migrate"),
+			},
+			objs:      []runtime.Object{depl, cm, kubeadmCM},
+			expectErr: true,
+		},
+		// TODO: Add test to return error if updateCoreDNSDeployment fails.
```

Comment:
If I'm not wrong also the happy case is not tested (CoreDNSDeployment is upgraded)